### PR TITLE
opam-0install-cudf: Fix conflict detection

### PR DESCRIPTION
**Pull request by kit-ty-kate**
While using [opam-health-check](https://github.com/kit-ty-kate/opam-health-check) I noticed an issue where both `ocaml-base-compiler` and `ocaml-variants` would be installed despite being marked is incompatible using the `conflict-class` field.

CUDF does not have conflict classes like opam does, so opam transform them to regular conflicts. However some conflicts are recursive in some cases, for instance:
```
package: ocaml-base-compiler
version: 42
depends: ocaml = 73 , base-unix , base-bigarray , base-threads
conflicts: ocaml-variants , ocaml-system , ocaml-base-compiler
installed: true
version-lag: 2
opam-name: ocaml-base-compiler
opam-version: 4.10.1

package: ocaml-variants
version: 507
depends: ocaml = 73 , base-unix , base-bigarray , base-threads
conflicts: ocaml-system , ocaml-base-compiler , ocaml-variants
installed: true
version-lag: 81
installed-root: true
opam-name: ocaml-variants
opam-version: 4.10.1+pr9812
```
So recursive conflicts have to be removed in order to be used by `0install-solver`. This is done here: https://github.com/kit-ty-kate/opam-0install-solver/commit/2a42f6b365c9cf7649bdae1b2db5b78770b9adbd#diff-6c67ec95f62e252765f5844c28d784d2R112

The rest of the issue was that conflicts need to be present to be reverted later. This is done here: https://github.com/kit-ty-kate/opam-0install-solver/commit/2a42f6b365c9cf7649bdae1b2db5b78770b9adbd#diff-6c67ec95f62e252765f5844c28d784d2R116-R119

And the last issue was that I forgot to actually implement rejections.. This is now done here: https://github.com/kit-ty-kate/opam-0install-solver/commit/2a42f6b365c9cf7649bdae1b2db5b78770b9adbd#diff-6c67ec95f62e252765f5844c28d784d2R132

This PR also reduces the differences between `lib-cudf/model.ml` and `lib/model.ml` to improve such debug in the future (it still took me more than a day to get it..).

The diff was optimized using/for `vimdiff`. If that can help someone, here are the settings I used to get vimdiff usable with the default OCaml highlights:
```vim
highlight DiffAdd cterm=none ctermfg=green ctermbg=black
highlight DiffDelete cterm=none ctermfg=darkred ctermbg=black
highlight DiffChange cterm=none ctermfg=none ctermbg=black
highlight DiffText cterm=none ctermfg=black ctermbg=darkyellow
```

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
I'm currently testing this on a larger scale but so far it seems to hold up fine.

**Review by talex5 — APPROVED** [review]
I don't know the details of the CUDF stuff, but it looks reasonable to me. It's strange though that opam is generating conflicts between a package and itself, and then expecting the solver to filter those out. Do we know why it does that?

**Comment by kit-ty-kate** [comment]
> It's strange though that opam is generating conflicts between a package and itself, and then expecting the solver to filter those out. Do we know why it does that?

@AltGr?

> I'm currently testing this on a larger scale but so far it seems to hold up fine.

I think I've tested enough this change locally. I have one more round of test using opam-health-check then I'll merge and make a release.

**Comment by kit-ty-kate** [comment]
I haven't noticed any issues in the past month. Merging now, I'll do a release right away.

**Comment by AltGr** [comment]
It's the CUDF standard that dictates it this way: if you don't declare a self-conflict, you allow co-existing installations of different versions of the same package.

**Comment by kit-ty-kate** [comment]
> It's the CUDF standard that dictates it this way: if you don't declare a self-conflict, you allow co-existing installations of different versions of the same package.

Of course, but I think the question was more: Why does opam not do something like:
```
Package: ocaml-base-compiler
Version: 42
Conflicts: ocaml-base-compiler (<> 42)
```
instead of conflicting with every versions of the package itself (including version 42 in this example) like it does currently.

**Comment by AltGr** [comment]
While that would indeed be correct, it would add weight and the CUDF standard explicitely advises a more general self-conflict without a version constraint.
A reason might be that the version constraint would lead the underlying solver to use a much more convoluted encoding for the conflicts, blowing up in size and complexity (e.g. I could imagine that instead of one LP constraint `v1 + v2 + ... + vn <= 1`, we would end up with one constraint for every version (`v1 ⇒ v2 + ... + vn = 0` etc.), or even n² in the worst case).
In other words, the pattern is simple enough this way for the underlying solvers to detect and optimise it, although indeed it looks weird at first.
